src/zjs_gpio.c

```diff
@@ -420,11 +490,398 @@ static ZJS_DECL_FUNC(zjs_gpio_open_async)
+ *
+ * @name wire
+ * @param {GPIOPin} pin1
+ * @param {GPIOPin} pin2
+ */
+static ZJS_DECL_FUNC(zjs_gpio_mock_wire)
+{
+    ZJS_VALIDATE_ARGS(Z_OBJECT, Z_OBJECT);
+
+    struct device *port1, *port2;
+    int pin1, pin2;
+    lookup_pin(argv[0], &port1, &pin1);
+    lookup_pin(argv[1], &port2, &pin2);
+
+    jerry_value_t port1_obj = DEVICE_TO_PORT(port1);
+    jerry_value_t name1 = find_property_with_value(mock_root_obj, port1_obj);
+
+    jerry_value_t port2_obj = DEVICE_TO_PORT(port2);
+    jerry_value_t name2 = find_property_with_value(mock_root_obj, port2_obj);
+
+    if (!name1 || !name2) {
+        return zjs_error("invalid port object");
+    }
+
+    ZVAL pin1_obj = get_pin(port1_obj, pin1);
+    ZVAL pin2_obj = get_pin(port2_obj, pin2);
+    if (!jerry_value_is_object(pin1_obj) ||
+        !jerry_value_is_object(pin2_obj)) {
+        return zjs_error("invalid pin object");
+    }
+
+    ZVAL pin1_wired = zjs_get_property(pin1_obj, "wired");
+    ZVAL new1_wired = push_array(pin1_wired, pin2_obj);
+    zjs_set_property(pin1_obj, "wired", new1_wired);
+
+    ZVAL pin2_wired = zjs_get_property(pin2_obj, "wired");
+    ZVAL new2_wired = push_array(pin2_wired, pin1_obj);
+    zjs_set_property(pin2_obj, "wired", new2_wired);
+
+    return ZJS_UNDEFINED;
+}
+
+// mocked apis
+static struct device *mock_device_get_binding(const char *name)
+{
+    ZVAL obj = zjs_get_property(mock_root_obj, name);
+    jerry_value_t rval = obj;
+    if (!jerry_value_is_object(obj)) {
+        ZVAL new_obj = jerry_create_object();
+        zjs_set_property(mock_root_obj, name, new_obj);
+        rval = new_obj;
+    }
+
+    // This is pretty evil, but the struct device * is an opaque pointer type,
+    //   so it's really just a unique ID for our mock purposes. This also means
+    //   we don't reference the object on return as with normal jerry_value_t
+    //   rules.
+    return PORT_TO_DEVICE(rval);
+}
+
+static int mock_gpio_pin_read(struct device *port, uint32_t pin,
+                              uint32_t *value)
+{
+    jerry_value_t port_obj = DEVICE_TO_PORT(port);
+    if (!find_property_with_value(mock_root_obj, port_obj)) {
+        ERR_PRINT("invalid port object\n");
+        return -1;
+    }
+
+    ZVAL pin_obj = get_pin(port_obj, pin);
+    if (!jerry_value_is_object(pin_obj)) {
+        ERR_PRINT("invalid pin object\n");
+        return -1;
+    }
+
+    // reading from a GPIO output seems to be allowed, so don't check direction
+
+    // check for a 'state'
+    bool flag;
+    if (zjs_obj_get_boolean(pin_obj, "state", &flag)) {
+        *value = flag ? 1 : 0;
+        return 0;
+    }
+
+    // check for a pin this is wired to
+    ZVAL wired = zjs_get_property(pin_obj, "wired");
+    if (jerry_value_is_array(wired)) {
+        uint32_t len = jerry_get_array_length(wired);
+        for (uint32_t i = 0; i < len; ++i) {
+            ZVAL pin = jerry_get_property_by_index(wired, i);
+            if (zjs_obj_get_boolean(pin, "state", &flag)) {
```

Comment:
It's invalid wiring if you have multiple pins driving different voltages; in that case you'd get current flowing and a voltage somewhere between "0" and "1", so it would be undefined basically. So basically, not our problem here.